### PR TITLE
📘 SFT doc rewrite

### PR DESCRIPTION
**Pull request by qgallouedec**

### CODE REVIEW

File: docs/source/liger_kernel_integration.md [hunk 1]
```diff
@@ -4,4 +4,29 @@
 
 Section under construction. Feel free to contribute!
 
-</Tip>
\ No newline at end of file
+</Tip>
+
+[Liger Kernel](https://github.com/linkedin/Liger-Kernel) is a collection of Triton kernels designed specifically for LLM training. It can effectively increase multi-GPU training throughput by 20% and reduce memory usage by 60%. That way, we can **4x** our context length, as described in the benchmark below. They have implemented Hugging Face Compatible `RMSNorm`, `RoPE`, `SwiGLU`, `CrossEntropy`, `FusedLinearCrossEntropy`, and more to come. The kernel works out of the box with [Flash Attention](https://github.com/Dao-AILab/flash-attention), [PyTorch FSDP](https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html), and [Microsoft DeepSpeed](https://github.com/microsoft/DeepSpeed).
```
Comment (kashif) [inline docs/source/liger_kernel_integration.md]:
```suggestion
[Liger Kernel](https://github.com/linkedin/Liger-Kernel) is a collection of Triton kernels designed specifically for LLM training. It can effectively increase multi-GPU training throughput by 20% and reduce memory usage by 60%. That way, we can **4x** our context length, as described in the benchmark below. They have implemented Hugging Face compatible `RMSNorm`, `RoPE`, `SwiGLU`, `CrossEntropy`, `FusedLinearCrossEntropy`, with more to come. The kernel works out of the box with [Flash Attention](https://github.com/Dao-AILab/flash-attention), [PyTorch FSDP](https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html), and [Microsoft DeepSpeed](https://github.com/microsoft/DeepSpeed).
```

File: docs/source/sft_trainer.md [hunk 2]
```diff
@@ -1,675 +1,240 @@
...
-    peft_config=peft_config
-)
-
-trainer.train()
+training_args = SFTConfig(packing=True)
 ```
 
-> [!WARNING]
-> If the chat template contains special tokens like `<|im_start|>` (ChatML) or `<|eot_id|>` (Llama), the embedding layer and LM head must be included in the trainable parameters via the `modules_to_save` argument. Without this, the fine-tuned model will produce unbounded or nonsensical generations. If the chat template doesn't contain special tokens (e.g., Alpaca), then the `modules_to_save` argument can be ignored or set to `None`.
+For more defails on packing, see [Packing](reducing_memory_usage#packing).
```
Comment (kashif) [inline docs/source/sft_trainer.md]:
```suggestion
For more details on packing, see [Packing](reducing_memory_usage#packing).
```

File: docs/source/sft_trainer.md [hunk 3]
```diff
@@ -1,675 +1,240 @@
...
 
+### Train on assistant messages only
 
-You can also continue training your `PeftModel`. For that, first load a `PeftModel` outside `SFTTrainer` and pass it directly to the trainer without the `peft_config` argument being passed.
-
-### Training adapters with base 8 bit models
-
-For that, you need to first load your 8 bit model outside the Trainer and pass a `PeftConfig` to the trainer. For example:
+To train on assistant messages only, use a [conversational](dataset_formats#conversational) dataset and set `assistant_only_loss=True` in the [`SFTConfig`]. This setting ensures that loss is computed **only** on the assistant responses, ignoring user and system and user messages.
```
Comment (kashif) [inline docs/source/sft_trainer.md]:
```suggestion
To train on assistant messages only, use a [conversational](dataset_formats#conversational) dataset and set `assistant_only_loss=True` in the [`SFTConfig`]. This setting ensures that loss is computed **only** on the assistant responses, ignoring user or system messages.
```

File: docs/source/sft_trainer.md [hunk 4]
```diff
@@ -1,675 +1,240 @@
...
-    trainer.train()
+training_args = SFTConfig(assistant_only_loss=True)
 ```
 
-Note that you cannot train your model using Flash Attention 1 on an arbitrary dataset as `torch.scaled_dot_product_attention` does not support training with padding tokens if you use Flash Attention kernels. Therefore, you can only use that feature with `packing=True`. If your dataset contains padding tokens, consider switching to Flash Attention 2 integration.
+![](https://huggingface.co/datasets/trl-lib/documentation-images/resolve/main/train_on_assistant.png)
 
-Below are some numbers you can get in terms of speedup and memory efficiency, using Flash Attention 1, on a single NVIDIA-T4 16GB.
+> [!WARNING]
+> This functionality is only available for chat templates that support returning the assistant tokens mask via the `{% raw %}{% generation %}{% endraw %}` keyword. For an example of such an template, see [Qwen/Qwen3-8B/discussions/14](https://huggingface.co/Qwen/Qwen3-8B/discussions/14).
```
Comment (kashif) [inline docs/source/sft_trainer.md]:
```suggestion
> This functionality is only available for chat templates that support returning the assistant tokens mask via the `{% raw %}{% generation %}{% endraw %}` keyword. For an example of such a template, see [Qwen/Qwen3-8B/discussions/14](https://huggingface.co/Qwen/Qwen3-8B/discussions/14).
```

File: docs/source/sft_trainer.md [hunk 5]
```diff
@@ -1,675 +1,240 @@
...
-)
-```
+Liger Kernel is a collection of Triton kernels for LLM training that boosts multi-GPU throughput by 20%, cuts memory use by 60% (enabling up to 4× longer context), and works seamlessly with tools like Flash Attention, PyTorch FSDP, and DeepSpeed. For more information, see [Liger Kernel Integration](liger_kernel_integration).
 
-If you don't use quantization, make sure your model is loaded in half-precision and dispatch your model on a supported GPU device.
-After loading your model, you can either train it as it is or attach adapters and train adapters on it in case your model is quantized.
+### Train with Unsloth
 
-In contrast to Flash Attention 1, the integration makes it possible to train your model on an arbitrary dataset that also includes padding tokens.
+Unsloth is an open‑source framework for fine‑tuning and reinforcement learning that trains LLMs (like Llama, Mistral, Gemma, DeepSeek, and more) up to 2× faster with up to 70% less VRAM, while providing a streamlined, Hugging Face–compatible workflow for training, evaluation, and deployment. Fo more information, see [Unsloth Integration](unsloth_integration).
```
Comment (kashif) [inline docs/source/sft_trainer.md]:
```suggestion
Unsloth is an open‑source framework for fine‑tuning and reinforcement learning that trains LLMs (like Llama, Mistral, Gemma, DeepSeek, and more) up to 2× faster with up to 70% less VRAM, while providing a streamlined, Hugging Face–compatible workflow for training, evaluation, and deployment. For more information, see [Unsloth Integration](unsloth_integration).
```

**Review by kashif — APPROVED** [review]

**Review by edbeeching — APPROVED** [review]

File: docs/source/sft_trainer.md [hunk 6]
```diff
@@ -1,568 +1,293 @@
...
-trainer.train()
+training_args = SFTConfig(assistant_only_loss=True)
 ```
 
-### Control over the pretrained model
+![train_on_assistant](https://huggingface.co/datasets/trl-lib/documentation-images/resolve/main/train_on_assistant.png)
 
-You can directly pass the kwargs of the `from_pretrained()` method to the [`SFTConfig`]. For example, if you want to load a model in a different precision, analogous to
+<Tip warning={true}>
+This functionality is only available for chat templates that support returning the assistant tokens mask via the `{% generation %}` and `{% endgeneration %}` keywords. For an example of such an template, see [HugggingFaceTB/SmolLM3-3B](https://huggingface.co/HuggingFaceTB/SmolLM3-3B/blob/main/chat_template.jinja#L76-L82).
```
Comment (kashif) [inline docs/source/sft_trainer.md]:
```suggestion
This functionality is only available for chat templates that support returning the assistant tokens mask via the `{% raw %}{% generation %}{% endraw %}` and `{% raw %}{% endgeneration %}{% endraw %}` keywords. For an example of such a template, see [HugggingFaceTB/SmolLM3-3B](https://huggingface.co/HuggingFaceTB/SmolLM3-3B/blob/main/chat_template.jinja#L76-L82).
```

Comment (kashif) [reply docs/source/sft_trainer.md]: need to put any `{% ... %}` inside `raw` and `endraw` for the doc to render else one gets an error
```
[vite-plugin-svelte] /tmp/tmpeuz180__/kit/src/routes/sft_trainer/+page.svelte:225:115 Unexpected token
file: /tmp/tmpeuz180__/kit/src/routes/sft_trainer/+page.svelte:225:115
 223 |  ></p>
 224 |  <Tip warning={true}>
 225 |  This functionality is only available for chat templates that support returning the assistant tokens mask via the `{% generation %}` and `{% endgeneration %}` keywords. For an example of such an template, see [HugggingFaceTB/SmolLM3-3B](https://huggingface.co/HuggingFaceTB/SmolLM3-3B/blob/main/chat_template.jinja#L76-L82).
 ```

File: docs/source/sft_trainer.md [hunk 7]
```diff
@@ -1,568 +1,293 @@
...
-Note that all keyword arguments of `from_pretrained()` are supported.
+<Tip>
+Training on completion only is compatible with training on assistant messages only. In this case, use a [conversational](dataset_formats#conversational) [prompt-completion](dataset_formats#prompt-completion) dataset and set `assistant_only_loss=True` in the [`SFTConfig`].
+</Tip>
 
-### Training adapters
+### Train adapters with PEFT
 
-We also support tight integration with 🤗 PEFT library so that any user can conveniently train adapters and share them on the Hub instead of training the entire model.
+We support tight integration with 🤗 PEFT library so that any user can conveniently train adapters and share them on the Hub instead of training the entire model.
```
Comment (kashif) [inline docs/source/sft_trainer.md]:
```suggestion
We support tight integration with 🤗 PEFT library, allowing any user to conveniently train adapters and share them on the Hub, rather than training the entire model.
```

File: docs/source/unsloth_integration.md [hunk 8]
```diff
@@ -4,4 +4,64 @@
 
 Section under construction. Feel free to contribute!
 
-</Tip>
\ No newline at end of file
+</Tip>
+
+Unsloth is an open‑source framework for fine‑tuning and reinforcement learning that trains LLMs (like Llama, Mistral, Gemma, DeepSeek, and more) up to 2× faster with up to 70% less VRAM, while providing a streamlined, Hugging Face–compatible workflow for training, evaluation, and deployment.
+Unsloth library that is fully compatible with [`SFTTrainer`]. Some benchmarks on 1x A100 listed below:
```
Comment (lewtun) [inline docs/source/unsloth_integration.md]: turbo nit:

```suggestion
Unsloth library that is fully compatible with [`SFTTrainer`]. Some benchmarks on 1 x A100 listed below:
```

File: docs/source/unsloth_integration.md [hunk 9]
```diff
@@ -4,4 +4,64 @@
...
+
+| 1 A100 40GB     | Dataset   | 🤗   | 🤗 + Flash Attention 2 | 🦥 Unsloth | 🦥 VRAM saved |
```
Comment (lewtun) [inline docs/source/unsloth_integration.md:12]: Is the `VRAM saved` relative to the 🤗 + FA2 combination or the vanilla 🤗 one? Might be good to specify for clarity

File: docs/source/unsloth_integration.md [hunk 10]
```diff
@@ -4,4 +4,64 @@
...
+
+max_length = 2048 # Supports automatic RoPE Scaling, so choose any number
+
+# Load model
+model, tokenizer = FastLanguageModel.from_pretrained(
+    model_name="unsloth/mistral-7b",
+    max_seq_length=max_length,
+    dtype=None,  # None for auto detection. Float16 for Tesla T4, V100, Bfloat16 for Ampere+
+    load_in_4bit=True,  # Use 4bit quantization to reduce memory usage. Can be False
+    # token = "hf_...", # use one if using gated models like meta-llama/Llama-2-7b-hf
```
Comment (lewtun) [inline docs/source/unsloth_integration.md]: We don't recommend passing tokens explicitly anymore, so I would remove this avoid user leaks (should be automatically handled by `huggingface-cli` now)

**Review by lewtun — APPROVED** [review]
LGTM with some nits

File: docs/source/sft_trainer.md (hunk 6 above)
Comment (qgallouedec) [reply docs/source/sft_trainer.md]: thanks!

File: docs/source/sft_trainer.md [hunk 11]
```diff
@@ -1,568 +1,295 @@
...
     train_dataset=dataset,
-    args=SFTConfig(output_dir="Qwen2.5-0.5B-SFT"),
-    peft_config=peft_config
+    peft_config=LoraConfig()
 )
 
 trainer.train()
 ```
 
-> [!WARNING]
```
Comment (romitjain) [inline docs/source/sft_trainer.md:274]: Hi @qgallouedec 

I am just wondering why this warning was removed? Is this something that has been resolved?

Comment (qgallouedec) [reply docs/source/sft_trainer.md:274]: Now it's handled automatically if you use clone_chat_template (or chat_template_path from SFTConfig)

Comment (romitjain) [reply docs/source/sft_trainer.md:274]: Thanks for the quick reply @qgallouedec but based on my reading of the documentation, it does not handle the `modules_to_save` issue that was described in this warning.

The user still needs to explicitly add `modules_to_save` in case new tokens are being added.

Is this the correct interpretation or am I missing something?

Comment (romitjain) [reply docs/source/sft_trainer.md:274]: @qgallouedec Sorry for pinging again - But hoping I could understand this. Can you please point me to the correct implementation?
